all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots for that task type at all
+  times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For example,
+  in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
+
+## Proposed interface for slot management
+
+To reiterate, we can't possibly know as well as the user what resources a given workflow or activity task will consume.
+Thus, we need to provide an interface by which users can tell us whether they think it's reasonable to allow the use
+of an additional slot.
+
+The interface definition here is provided in Rust, but need not be copied exactly by other languages. Since users are
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error>;
+
+  /// Tries to immediately reserve a slot, returning true if a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  fn try_reserve_slot(&self) -> Result<bool, anyhow::Error>;
+
+  /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
+  /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
+  /// then the slot can actually be used to work on the task.
+  /// 
+  /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
+  /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
+  fn mark_slot_used(&self, info: &Self::SlotKind::Info);
```

Comment:
I think it is an important distinction even the basic `FixedSlotSupplier` or `ResourceBoundSlotSupplier` need this information. `FixedSlotSupplier` for metrics and `ResourceBoundSlotSupplier` can use it for part of its logic. If we want to support passing any information on the reserved slot then the distinction is important.

Reply:
> FixedSlotSupplier for metrics
Which metrics? If we're talking about new metrics, we can give their definitions (and I consider a "used" slot to be a reserved one in that case).

> ResourceBoundSlotSupplier can use it for part of its logic

I don't believe it needs to. I think this complicates the interface and logic. Having said that, I don't mind it that much.

Reply:
>which metrics

`worker_task_slots_available`

I think it is important for `SlotSuppliers` to have the ability to know if a slot is just reserved vs actually being used.  Consider also a slot can be reserved, but never used. I don't think it complicates the interface, and if an implementation doesn't need to distinguish they just have a noop implementation so i don't see how it complicates the logic. In langs that support it we can provide a default noop implementation if we want.

Reply:
`worker_task_slots_available` is unused slots. I think that is the correct definition since pollers reserve a slot before making the poll request so there may be no more unreserved slots, but the worker can still accept `unused slot` amount of work. Eager stuff does make it awkward I agree.

Reply:
> worker_task_slots_available is unused slots [...] Eager stuff does make it awkward I agree.

More than awkward, it makes it wrong. We may have to change the definition of "used" to "started a task or requested eager" instead of just "started a task" to make it be compatible. I am ok w/ that, but we should probably make sure to `mark_slot_used` as soon as we get a slot for eager use then.

Reply:
I don't think it is wrong, I don't think there is a material difference between eager requests and polling a task queue. Those eager requests are just asking for a new task with a filter for what types they accept. I don't think we need to change any existing metric definition here. I would support adding a new metric to capture the unserved slots or capture the number of "pending"

Reply:
So I am advocating to maintain compatibility 

>Today if, before I make a wft completed call to server, I reserve a slot for eager activity, does worker_task_slots_available show it as no longer available?

worker_task_slots_available  would show as available

Reply:
> worker_task_slots_available would show as available

Ok, my mistake. I thought the simple act of reserving for eager today affected the metric. Then nevermind, I agree then reservation and usage can remain separate. (can consider thread about this resolved)